### PR TITLE
BACK-1589 Fix large account synchronization

### PR DESCRIPTION
**Pull request by mbasire-ledger**
## What is this about?

Limit the usage of memory resources by the Worker

  * Accumulate BatchResult by addresses range in order to detect empty
pages
  * Max block height carry by BatchResult instead of the whole list of
Transactions
  * Bookkeeper hides its internal stream
  * Txs of batch addresses streamed to interpreter (the size of the
batch may be a problem but explorer bug expect us to use a significant
amount of tx by batch)

## Cute picture of lama
lama

### CODE REVIEW

File: coins/bitcoin/worker/src/main/scala/co/ledger/lama/bitcoin/worker/models/BatchResult.scala [hunk 1]
```diff
@@ -1,9 +1,26 @@
...
+
+  implicit def monoid[A <: Transaction]: Monoid[BatchResult[A]] = new Monoid[BatchResult[A]] {
+    override def empty: BatchResult[A] = BatchResult(List.empty, None)
+
+    override def combine(x: BatchResult[A], y: BatchResult[A]): BatchResult[A] =
+      BatchResult(
+        addresses = (x.addresses ::: y.addresses).distinct,
+        maxBlock = {
+          val blocks = List(x.maxBlock, y.maxBlock).flatten
+          if (blocks.nonEmpty) Some(blocks.max) else None
```
Comment (twilgenbus-ledger) [inline coins/bitcoin/worker/src/main/scala/co/ledger/lama/bitcoin/worker/models/BatchResult.scala:22]: Could use `maxOption`

**Review by twilgenbus-ledger — CHANGES_REQUESTED** [review]
LGTM

Non functionnal changes

File: coins/bitcoin/worker/src/main/scala/co/ledger/lama/bitcoin/worker/models/BatchResult.scala (hunk 1 above)
Comment (mbasire-ledger) [reply coins/bitcoin/worker/src/main/scala/co/ledger/lama/bitcoin/worker/models/BatchResult.scala:22]: done

File: coins/bitcoin/worker/src/main/scala/co/ledger/lama/bitcoin/worker/models/BatchResult.scala [hunk 2]
```diff
@@ -1,9 +1,26 @@
 package co.ledger.lama.bitcoin.worker.models
 
-import co.ledger.lama.bitcoin.common.models.explorer.Transaction
+import cats.kernel.Monoid
+import co.ledger.lama.bitcoin.common.models.explorer.{Block, Transaction}
 import co.ledger.lama.bitcoin.common.models.interpreter.AccountAddress
 
 case class BatchResult[Tx <: Transaction](
```
[the remark this answers is not in the document]
Comment (mbasire-ledger) [reply coins/bitcoin/worker/src/main/scala/co/ledger/lama/bitcoin/worker/models/BatchResult.scala:7]: done. BatchResult moved to Bookkeeper.BatchResult

File: coins/bitcoin/worker/src/main/scala/co/ledger/lama/bitcoin/worker/services/Bookkeeper.scala [hunk 3]
```diff
@@ -11,148 +12,180 @@ import co.ledger.lama.bitcoin.common.models.explorer.{
...
+
+        val blocks = txs.collect { case ConfirmedTransaction(_, _, _, _, _, _, _, block, _) =>
+          block
+        }
+
+        save(interpreter, accountId, txs)
+          .as(
+            BatchResult(
+              usedAddresses,
+              if (blocks.isEmpty) None else Some(blocks.max)
```
[the remark this answers is not in the document]
Comment (mbasire-ledger) [reply coins/bitcoin/worker/src/main/scala/co/ledger/lama/bitcoin/worker/services/Bookkeeper.scala:107]: done

File: coins/bitcoin/worker/src/main/scala/co/ledger/lama/bitcoin/worker/services/Bookkeeper.scala [hunk 4]
```diff
@@ -11,148 +12,180 @@ import co.ledger.lama.bitcoin.common.models.explorer.{
...
+      addresses: Set[AccountAddress],
+      block: Option[BlockHash]
+  )(implicit recordable: Recordable[Tx]): Stream[IO, Tx] =
+    recordable.fetch(explorer)(addresses.map(_.accountAddress), block)
+
+  def save[Tx <: Transaction](
+      interpreter: InterpreterClient,
+      accountId: AccountId,
+      txs: List[Tx]
+  )(implicit recordable: Recordable[Tx]): IO[String] = recordable.save(interpreter)(accountId, txs)
```
[the remark this answers is not in the document]
Comment (mbasire-ledger) [reply coins/bitcoin/worker/src/main/scala/co/ledger/lama/bitcoin/worker/services/Bookkeeper.scala:124]: of course.

**Review by twilgenbus-ledger — APPROVED** [review]

**Pull request merged**